### ISSUE TITLE
`IncrementalTaskInputs` provide misleading information when working on ordered collections

### ISSUE THREAD

**Issue by melix** [gradle/gradle#1931]
Tasks which use on `IncrementalTaskInputs` cannot trust the list of additions/removals/changes reported by this class, when the underlying collection is _ordered_. This is in particular the case for inputs annotated with `@Claspath` or `@CompileClasspath`. Whenever and elements is added or removed, the information reported is totally wrong: files will be reported as updated when they are not, others will be reported as deleted when they are not, and deleted files will either be missing or reported wrongly as updated or added.

### Expected Behavior

`IncrementalTaskInputs` should report which files have changed propertly when the collection is ordered.

### Current Behavior

This bug is responsible for 2 problems which have been reported independently: #1761 and #1913. Those bugs are related to the incremental compiler, which uses `IncrementalTaskInputs`. The incremental compiler has been updated to mitigate this bug, and don't trust the inputs.

### Context

A proper fix might have consequences on performance, as described by @wolfs : 

> For `ORDERED` we are comparing each element in order - and if it matches it is no change, if it changed then we have a change. We try not to match sublists.
> In order to get something you expect I guess we would need to calculate something similar to the levenshtein distance, perhaps using this this algorithm: 
> https://en.wikipedia.org/wiki/Hirschberg%27s_algorithm
> The complexity of this would be O(n*m), where n and m are the number of elements of both file collection snapshots. I guess this would be too expensive.

### Steps to Reproduce (for bugs)

Add this to the `org.gradle.api.internal.changedetection.state.TaskFilePropertyCompareStrategyTest` test case:

```groovy
    @Unroll
    def "non-trivial removal with absolute paths with more than 2 files (#strategy)"() {
        expect:
        changesUsingAbsolutePaths(strategy,
            ["jcip-annotations.jar": snapshot("two"), "slf4j-api": snapshot("three")],
            ["commons-lang3.jar": snapshot("one"), "jcip-annotations.jar": snapshot("two"), "slf4j-api": snapshot("three")]
        ) == [change("commons-lang3.jar", REMOVED)]

        where:
        strategy << [ORDERED, UNORDERED, OUTPUT]
    }
```

It will fail for the `ORDERED` case.

**Comment by melix** [comment]
Here's an example of what can be seen when a class file is removed from classpath (`Utils2.class`):

```
Task :library:compileTestJava actions class loader hashes: [8fbf7bca23e9d538f445716c557d8f23, 5ade796f7ed2b45ef0f5b9dff90ef315]
Executing task ':library:compileTestJava' (up-to-date check took 0.014 secs) due to:
  Input property 'classpath' file /tmp/compile-avoidance/library/build/classes/java/main/org/gradle/sample/Utils2.class has been removed.
  Input property 'classpath' file /home/cchampeau/.gradle/caches/modules-2/files-2.1/org.apache.commons/commons-math3/3.6.1/e4ba98f1d4b3c80ec46392f25e094a6a2e58fcbf/commons-math3-3.6.1.jar has been added.
  Input property 'classpath' file /home/cchampeau/.gradle/caches/modules-2/files-2.1/com.google.guava/guava/21.0/3a3d111be1be1b745edfa7d91678a12d7ed38709/guava-21.0.jar has changed.
```

The incremental compiler gets notified that `Utils2.class` has been removed, but also that `commons-math3-3.6.1.jar` has been added (which is wrong) and that `guava-21.0.jar` has changed (which is also wrong). The [PR](https://github.com/gradle/gradle/pull/1930) works around this problem, but it nonetheless exists.

**Comment by alexey-tsvetkov** [comment]
I think this issue should be mentioned in the docs with a big warning sign.

**Comment by oehme** [comment]
Yes, this is indeed a serious limitation. @melix can you point @alexey-tsvetkov at how we mitigated this for incremental Java compilation? Somehow we must have worked around this, since it works for the Java case.

**Comment by melix** [comment]
Details are in the [linked pull request](https://github.com/gradle/gradle/pull/2094). Basically the incremental compiler ignores the inputs and computes its own alignment. This strategy could be used for incremental task inputs too, but we didn't proceed (higher priorities) given that there could be an impact on performance.

**Comment by wolfs** [comment]
Fixed via #8499.
